[ADC+RR] Random rotation feature

finnroblin opened the pull request:
### Description
Part of #2714 to add random rotation to k-NN for binary quantized indices. 

Random rotation can improve recall on some datasets by smoothing uneven variance across dimensions. See the [issue](https://github.com/opensearch-project/k-NN/issues/2714) for more details and benchmarks. 

This PR implements random rotation with an orthonormal rotation matrix. Each entry is sampled from a standard Gaussian variable and then the matrix is made orthonormal through Gram-Schmidt. The rotation process could be sped up with the panama API but that is out of scope for this PR. 

Random rotation can be enabled through an encoder-level index mapping:
```
PUT vector-index
{
  "settings" : {
    "index": {
      "knn": true
    }
  },
  "mappings": {
    "properties": {
      "vector_field": {
        "type": "knn_vector",
        "dimension": 8,
        "method": {
            "name": "hnsw",
            "engine": "faiss",
            "space_type": "l2",
            "parameters": {
              "encoder": {
                "name": "binary",
                "parameters": {
                  "bits": 1,
                  // NEW MAPPING PARAMETER
		  "enable_random_rotation": true
                }
              }
            }
        }
      }
    }
  }
}
```

Since the focus of the overlying issue is on binary quantization, random rotation is implemented only for binary indices. The rotation matrix is handled by the `ScalarQuantization` part of the quantization framework. 

- As described above, `QFrameBitEncoder` takes the new random rotation setting.
- `TrainingRequest` handles `isEnableRandomRotation`.
- `OneBitScalarQuantizationState`  and `MultiBitScalarQuantizationState` store the matrix if it exists. 
- Matrix generation is handled within the `QuantizationHelper.calculateQuantizationState` method. Once a matrix is generated and written to quantization state it does not change. 
- Rotation itself is handled by `XBitScalarQuantizer.quantize()`. If rotation is enabled then vectors are rotated before they are quantized. The thresholds used for quantization are (as a side effect) also rotated.
- The `aboveThresholdMeans`/`belowThresholdMeans` in scalar quantization states will be used in a follow-up PR to transform the vector for ADC (see the issue for more information).

### Related Issues
https://github.com/opensearch-project/k-NN/issues/2714

### Check List
- [X] New functionality includes testing.
- [ ] New functionality has been documented. [Will track documentation through RFC #2714].
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md). [No API changes]
- [X] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose). [Will track documentation through RFC #2714].

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

## Coauthors
Vikash Tiwari (@Vikasht34), Tal Wagner (@talwagner)

jmazanec15 commented on src/main/java/org/opensearch/knn/common/KNNConstants.java:
```diff
@@ -84,6 +84,9 @@ public class KNNConstants {
     public static final String RADIAL_SEARCH_KEY = "radial_search";
     public static final String MODEL_VERSION = "model_version";
     public static final String QUANTIZATION_STATE_FILE_SUFFIX = "osknnqstate";
+    public static final int QUANTIZATION_RANDOM_ROTATION_DEFAULT_SEED = 1212121212; // used to seed the RNG for reproducability in unit
```
Just to confirm, it will always be same rotation matrix for a given dimension size?

finnroblin replied:
That's the current implementation but I'm open to changing it. I don't think there's anything special about this particular seed -- Yonatan's experiments used this particular number so I moved it into the code.

The fixed seed gives reproducible results since the matrix is the same across multiple ingestions of the same data. I'm worried about a case where several otherwise equivalent benchmark runs produce different recall numbers due to different random matrices.

But as you pointed out it won't be truly random since the rotation matrices will be the same when we fix dimension size.

I think this is okay. Each gaussian in the matrix is still independent when we fix the seed. By central limit theorem we'll still get the property of random rotations that variance is smoothed across dimensions.

jmazanec15 commented on src/main/java/org/opensearch/knn/index/engine/faiss/QFrameBitEncoder.java:
```diff
@@ -41,7 +43,8 @@ public class QFrameBitEncoder implements Encoder {
      *     "parameters": {
-     *       "bits": 2
+     *       "bits": 2,
+     *       "enable_random_rotation": true,
```
I think just call it random_rotation. enable_random_rotation seems too long

jmazanec15 commented on src/main/java/org/opensearch/knn/index/engine/qframe/QuantizationConfigParser.java:
```diff
@@ -34,7 +35,17 @@ public static String toCsv(QuantizationConfig quantizationConfig) {
         }
 
-        return TYPE_NAME + SEPARATOR + BINARY_TYPE + "," + BIT_COUNT_NAME + SEPARATOR + quantizationConfig.getQuantizationType().getId();
+        return TYPE_NAME
```
Are we handling backwards compatibility here?

jmazanec15 commented on src/main/java/org/opensearch/knn/quantization/models/quantizationParams/ScalarQuantizationParams.java:
```diff
@@ -25,6 +26,11 @@
 @EqualsAndHashCode
 public class ScalarQuantizationParams implements QuantizationParams {
     private ScalarQuantizationType sqType;
+    private boolean isEnableRandomRotation = QFrameBitEncoder.DEFAULT_ENABLE_RANDOM_ROTATION;
```
nit: final? Also, do we use this somewhere?

finnroblin replied:
Good catch, I moved the `isEnableRandomRotation` to training request and didn't delete. I don't think scalarquantizationparams needs to store this and we can keep solely in training request and in the quantization state. Will remove.

finnroblin replied:
Took another look and `isEnableRandomRotation` is necessary for `QuantizationService::train` when that method constructs the training request.

If I make this field final I'll have to drop the `@AllArgsConstructor` and force the `@NoArgsConstructor`. What do you think?

jmazanec15 commented on src/main/java/org/opensearch/knn/quantization/models/quantizationState/MultiBitScalarQuantizationState.java:
```diff
@@ -73,12 +112,31 @@ public void writeTo(StreamOutput out) throws IOException {
      */
     public MultiBitScalarQuantizationState(StreamInput in) throws IOException {
-        int version = in.readVInt(); // Read the version
+        int version = in.readVInt();
```
Can we add back comments?

jmazanec15 commented on src/main/java/org/opensearch/knn/quantization/models/quantizationState/MultiBitScalarQuantizationState.java:
```diff
@@ -73,12 +112,31 @@ public void writeTo(StreamOutput out) throws IOException {
+        }
+
+        if (Version.fromId(version).onOrAfter(Version.V_3_1_0)) {
+            FloatArrayWrapper[] wrappedBelowMeans = in.readOptionalArray(FloatArrayWrapper::new, FloatArrayWrapper[]::new);
```
Are we using below or above means for random rotation? If not, could we remove and add them in adc change?

jmazanec15 commented on src/main/java/org/opensearch/knn/quantization/models/quantizationState/MultiBitScalarQuantizationState.java:
```diff
@@ -175,6 +233,39 @@ public long ramBytesUsed() {
+            this.array = in.readFloatArray();
+        }
+
+        public float[] getArray() {
```
nit: lombok getter

jmazanec15 commented on src/main/java/org/opensearch/knn/quantization/models/quantizationState/MultiBitScalarQuantizationState.java:
```diff
@@ -175,6 +233,39 @@ public long ramBytesUsed() {
+    private static class FloatArrayWrapper implements Writeable {
+        private final float[] array;
+
+        public FloatArrayWrapper(float[] array) {
```
nit: all arg ctor

jmazanec15 commented on src/main/java/org/opensearch/knn/quantization/models/quantizationState/OneBitScalarQuantizationState.java:
```diff
@@ -139,6 +211,41 @@ public long ramBytesUsed() {
         return size;
     }
+
+    private class FloatArrayWrapper implements Writeable {
```
Can we put in a util since its used in multiple places?

jmazanec15 commented on src/main/java/org/opensearch/knn/quantization/models/requests/TrainingRequest.java:
```diff
@@ -30,4 +42,6 @@ public abstract class TrainingRequest<T> {
      */
     public abstract T getVectorAtThePosition(int position) throws IOException;
+
+    public abstract void resetVectorValues();
```
javadoc

jmazanec15 commented on src/main/java/org/opensearch/knn/quantization/quantizer/QuantizerHelper.java:
```diff
@@ -5,104 +5,377 @@
+@Log4j2
 @UtilityClass
 class QuantizerHelper {
+    // ========================= ONE BIT ========================= //
```
nit: remove

jmazanec15 commented on src/main/java/org/opensearch/knn/quantization/quantizer/QuantizerHelper.java:
```diff
@@ -5,104 +5,377 @@
- * consistent handling of training requests and sampled indices.
+ * Utility class for calculating quantization state information for both
+ * OneBit and MultiBit scalar quantizers. Handles computing thresholds,
+ * below/above mean statistics, L2/L1 ratios, and rotation matrix application.
```
do we need l2/l1 ratios?

jmazanec15 commented on src/main/java/org/opensearch/knn/quantization/quantizer/QuantizerHelper.java:
```diff
@@ -5,104 +5,377 @@
+        ScalarQuantizationParams quantizationParams
+    ) throws IOException {
+        validateSampledIndices(sampledIndices);
+        int first_vec_idx = sampledIndices[0];
```
are we able to re-use more code between this and multi?

jmazanec15 commented on src/main/java/org/opensearch/knn/quantization/quantizer/RandomGaussianRotation.java:
```diff
@@ -0,0 +1,91 @@
+public class RandomGaussianRotation {
+
+    /**
+     * Generates a random rotation matrix using Gaussian distribution and orthogonalization.
```
Maybe add comment about preservation of dot products and l2 distances when this is applied. Also add comment about seeding

jmazanec15 commented on src/main/java/org/opensearch/knn/quantization/quantizer/RandomGaussianRotation.java:
```diff
@@ -0,0 +1,91 @@
+     */
+    public float[] applyRotation(float[] vector, float[][] rotationMatrix) {
+        int dimensions = vector.length;
+        float[] rotatedVector = new float[dimensions];
```
can this be in place instead of allocating a new vector?

finnroblin replied:
Do you mean make the method `void` and modify the vector in place? We need to make a vector copy either way or else the matrix-vector multiplication will break. At that point might as well return the modified vector and have one fewer side effect/no need to worry about java "modifying a reference" semantics.

jmazanec15 commented on src/main/java/org/opensearch/knn/training/FloatTrainingDataConsumer.java:
```diff
@@ -98,6 +98,12 @@ private List<byte[]> quantizeVectors(List<?> vectors) throws IOException {
+
+            @Override
+            public void resetVectorValues() {
+                // No-op
```
Why no-op?

finnroblin replied:
Since `getVectorAtThePosition` doesn't change the state of the vectors. `resetVectorValues()` is in the interface since for `KNNVectorQuantizationTrainingRequest` we need `resetVectorValues()` to reset the `knnVectorValuesSupplier` iterator state.

I'll add a clarifying comment here.

finnroblin commented:
> > I'm seeing a drop in recall for product quantization with these changes (testRecall_whenFaissHNSWPQFP32_thenRecallAbove50percent). Investigating whether erroneous trigger of random rotation is the cause.
> 
> It Should be for PQ use case right , why PQ Use call BQ Random rotation? Strange

Yeah, I'm digging into it right now. None of my rotation breakpoints are being triggered for PQ. 

> > I'm seeing a drop in recall for product quantization with these changes (testRecall_whenFaissHNSWPQFP32_thenRecallAbove50percent). Investigating whether erroneous trigger of random rotation is the cause.
> 
> It Should be for PQ use case right , why PQ Use call BQ Random rotation? Strange

Yes, my hunch is there's some side effect to the new `random_rotation` boolean in the FloatTrainingDataConsumer (even if it's always false for PQ and the vectors are never rotated). Currently debugging

finnroblin commented:
@Vikasht34 That one recall tests was failing since I changed the dimension in the RecallIT to 128 from 32 while testing and forgot to change back. Fixed now and verified that random rotation param is always off for HNSW PQ + IVF PQ methods.

finnroblin commented:
The integration test `org.opensearch.knn.index.SegmentReplicationIT.testSearchOnReplicas_whenIndexHasDeletedDocs_thenSuccess` is failing. It doesn't relate to this PR and it passes when I modify `thread.sleep` duration from `5000` to `1500`. I suspect the test is flaky.

The error is:
```
SegmentReplicationIT > testSearchOnReplicas_whenIndexHasDeletedDocs_thenSuccess FAILED
    java.lang.AssertionError: expected:<5> but was:<4>
        at __randomizedtesting.SeedInfo.seed([6C7B80E82ACDA81D:D01CF5F37EA2C41B]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:647)
        at org.junit.Assert.assertEquals(Assert.java:633)
        at org.opensearch.knn.index.SegmentReplicationIT.testSearchOnReplicas_whenIndexHasDeletedDocs_thenSuccess(SegmentReplicationIT.java:70)
```

I am assuming the test is flaky since modifying the `Thread.sleep` duration changes the behavior of the test. Locally I was able to get the test to pass by setting `Thread.sleep(1000)` instead of `Thread.sleep(5000)`. 

If I increase the sleep to 20 seconds then I get a different assertion error (3 docs in result as opposed to 4):
```
SegmentReplicationIT > testSearchOnReplicas_whenIndexHasDeletedDocs_thenSuccess FAILED
    java.lang.AssertionError: expected:<5> but was:<3>
        at __randomizedtesting.SeedInfo.seed([2C9EF4B6DE51F802:90F981AD8A3E9404]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:647)
        at org.junit.Assert.assertEquals(Assert.java:633)
        at org.opensearch.knn.index.SegmentReplicationIT.testSearchOnReplicas_whenIndexHasDeletedDocs_thenSuccess(SegmentReplicationIT.java:70)
```

finnroblin commented on src/main/java/org/opensearch/knn/index/engine/qframe/QuantizationConfigParser.java:
```diff
@@ -33,8 +35,19 @@ public static String toCsv(QuantizationConfig quantizationConfig) {
+            + quantizationConfig.getQuantizationType().getId();
 
-        return TYPE_NAME + SEPARATOR + BINARY_TYPE + "," + BIT_COUNT_NAME + SEPARATOR + quantizationConfig.getQuantizationType().getId();
+        if (Version.CURRENT.onOrAfter(Version.V_3_1_0)) {
```
Should use Lucene version for consistency with `fromCsv` method.

finnroblin commented on src/main/java/org/opensearch/knn/quantization/quantizer/RandomGaussianRotation.java:
```diff
@@ -0,0 +1,101 @@
+     * @param rotationMatrix The rotation matrix.
+     * @return The copy of the original vector but rotated.
+     */
+    public float[] applyRotation(float[] vector, float[][] rotationMatrix) {
```
I think batching will be challenging. We'd have to take a larger initiative to batch calls to `quantize`, which is probably a good idea but out of scope for this PR. I added calls to Lucene VectorUtils where applicable to take advantage of its SIMD capabilities.
